### PR TITLE
feat(mcp): session-scoped layer workspaces + ownership checks

### LINKED ISSUE #1030: layers/mcp: session-scoped workspaces + ownership checks for multi-client deployments (L5)
Tracking issue for a deliberate scope cut in #1027.

## Current state

`packages/mcp/src/tools/layer-store.ts` keeps one in-memory layer workspace (`drafts`, `layers`, `refs`, `reviews`) per **process**. This is correct for the registry-less local mode: the stdio MCP transport is one client per process, so process scope *is* session scope. The constraint is documented at the `active` declaration.

## What's deferred

Before the layer tools can be exposed over a multi-client transport (Streamable HTTP) or by the hosted registry (phase L5, `docs/architecture/layer-prs/12-roadmap.md`):

- [ ] Session identity in `ToolContext` (currently none exists to key on)
- [ ] Workspace storage keyed by session/client id (`getLayerWorkspace(sessionId)`)
- [ ] Drafts/reviews bound to an owning principal, with ownership checks on every mutating tool
- [ ] Error details audited so unknown-id responses don't aid cross-tenant enumeration

Raised by CodeRabbit review on #1027 (packages/mcp/src/tools/layer-store.ts discussion).

### PR DESCRIPTION
**Pull request by louistrue**
Closes #1030. Stacked on #1027 (targets its branch; re-target to `main` after #1027 merges).

## What

Implements all four deferred items from the issue so the layer tools are safe to expose over multi-client transports:

1. **Session identity in `ToolContext`** — new `SessionIdentity { id?, principal? }`; `id` is the transport session id (the HTTP transport already generates one per session and passes it to `SessionFactory.build`, now forwarded into `createMCPServer` via a new `MCPServerOptions.sessionId`), `principal` is `AuthScope.user` read at call time so `setScope` token refreshes apply. **`HttpTransport` rejects session creation when a custom `SessionFactory` builds the server without binding the provided session id** — otherwise every HTTP session would silently share the local workspace.
2. **Storage split** — `getLayerWorkspace(sessionId?)`: **drafts are session-private** (they're live CRDT sessions; destroyed via `disposeLayerWorkspace` on session end through `MCPServer.detach()`, never on SSE client drops), while **published layers, refs, and reviews are process-shared** — they're the collaboration surface, and over HTTP each principal necessarily holds its own session (the transport rejects scope mismatches), so reviews keyed to a session would be unreachable by their reviewers. stdio/in-process keeps the local draft space — registry-less local mode unchanged.
3. **Ownership** — drafts and reviews record the creating principal. Access matrix (documented at the helpers): `draft_apply_ops`/`publish_layer` draft-owner only; reviews visible to owner + listed reviewers **including reads** (`get_review_feedback`), since the review store spans principals; `respond_to_review` owner-only; `add_review_feedback` owner or listed reviewer. Unowned records (local/unauthenticated) stay accessible.
4. **Enumeration audit** — ownership/visibility denials are byte-identical to unknown-id errors, and every `ENTITY_NOT_FOUND` `details` list (`drafts`, `reviews`) only enumerates ids visible to the caller. Published layers stay fully listed (immutable, shared by design).

## Review feedback addressed

- **P1 (reviews unreachable across sessions)** → resolved by the storage split above; new `cross-session review collaboration` tests cover the exact flow: Alice publishes + requests review from her session, Bob acts on it from his, a stranger probing the id sees a nonexistent review, and Alice responds from a *later* reconnect session while follow-up drafts stay session-private.
- **P1 (optional `sessionId` in custom factories)** → `HttpTransport` now fails session creation with a 500 + explicit configuration error unless `server.sessionId` matches the id it handed the factory; `SessionFactory` contract documented.

## Tests

`layer-session.test.ts` (12 tests): cross-session draft isolation incl. no id leakage in errors, local-vs-session separation, dispose → fresh draft space while published layers/refs survive, owner-denial indistinguishable from unknown id, details filtering with mixed principals, review owner/reviewer/stranger matrix for reads and writes, cross-session review loop. `@ifc-lite/mcp`: 10 files / 72 tests pass, `tsc --noEmit` clean.

### CODE REVIEW

File: packages/mcp/src/tools/layer-store.ts [hunk 1]
```diff
@@ -81,22 +88,44 @@ export function createLayerWorkspace(): LayerWorkspace {
...
+const workspaces = new Map<string, LayerWorkspace>();
 
-export function getLayerWorkspace(): LayerWorkspace {
-  return active;
+/** Map key for transports without a session id (stdio, in-process, tests). */
+const LOCAL_WORKSPACE_KEY = 'local';
+
+export function getLayerWorkspace(sessionId?: string): LayerWorkspace {
+  const key = sessionId ?? LOCAL_WORKSPACE_KEY;
+  let ws = workspaces.get(key);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/mcp/src/tools/layer-store.ts]: Fixed in 9888d948 by splitting the store: drafts stay session-private (live CRDT docs, disposed with the session), while published layers, refs, and reviews are process-shared — so Bob reaches Alice's review from his own session. Review visibility now includes listed reviewers and gates reads too (the review store spans principals); denials remain byte-identical to unknown-id errors. New `cross-session review collaboration` tests cover the request_review → add_review_feedback → respond_to_review loop across three sessions, including an owner reconnecting under a fresh session id.

File: packages/mcp/src/server.ts [hunk 2]
```diff
@@ -93,6 +94,12 @@ export interface MCPServerOptions {
   capabilities?: ServerCapabilities;
   /** Optional pre-built viewer manager. Default: a fresh one bound to the registry. */
   viewer?: ViewerManager;
+  /**
+   * Transport session id for multi-client transports (one MCPServer per
+   * HTTP session). Keys per-session state like the layer workspace (#1030);
+   * omit for stdio/in-process where the process is the session.
+   */
+  sessionId?: string;
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/mcp/src/server.ts:102]: Fixed in 9888d948: `HttpTransport` now refuses session creation (500 + explicit configuration error) when `server.sessionId` doesn't match the id it handed the factory, so a custom `SessionFactory` that drops the second `build` argument fails loudly instead of silently sharing the local workspace. The `SessionFactory` contract documents the requirement.

**Comment by louistrue** [comment]
@codex review

File: packages/mcp/src/server.ts [hunk 3]
```diff
@@ -155,6 +164,11 @@ export class MCPServer {
     // The viewer holds an HTTP server + SSE listener; closing it stops
     // dangling sockets when the transport disconnects.
     if (this.viewer.isOpen()) this.viewer.close();
+    // Per-session layer drafts hold live Y.Docs — free them with the
+    // session. detach() only fires on true termination (HTTP DELETE /
+    // transport close), never on SSE client drops, so a reconnecting
+    // client keeps its workspace until the session actually ends.
+    if (this.sessionId !== undefined) disposeLayerWorkspace(this.sessionId);
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/mcp/src/server.ts:171]: Fixed in 6418a656: DELETE now enforces the same exact-scope-identity rule POST applies (403 on mismatch, nothing disposed), and GET/SSE attach got the same gate — a leaked session id must not let a foreign token read the victim's event stream either. Added transport-level tests covering DELETE/GET/POST identity plus the session-factory binding rejection.

**Comment by louistrue** [comment]
@codex review

File: packages/mcp/src/transport/http.ts [hunk 4]
```diff
@@ -215,6 +249,17 @@ export class HttpTransport {
       }
       const newId = randomUUID();
       const server = await this.opts.sessionFactory.build(scope, newId);
+      // A factory that drops the session id would put every HTTP session
+      // on the shared local layer workspace (cross-session reads/writes,
+      // no disposal) — refuse the deployment bug instead of running unsafe.
+      if (server.sessionId !== newId) {
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/mcp/src/transport/http.ts:255]: Fixed in 8bb0305d — good catch: the `ifc-lite mcp` command's factory in `packages/cli/src/commands/mcp.ts` predated the option and dropped the second `build` argument; it now forwards `sessionId`. Repo-wide search confirms these two (`packages/mcp/src/cli.ts`, `packages/cli/src/commands/mcp.ts`) are the only in-repo `SessionFactory` implementations, and both bind the id. Changeset now also bumps `@ifc-lite/cli`.

**Comment by louistrue** [comment]
@codex review

File: packages/mcp/src/tools/layer-store.ts [hunk 5]
```diff
@@ -294,32 +344,106 @@ export function createDraft(ws: LayerWorkspace, init: CreateDraftInit): DraftLay
...
 }
 
-export function requireDraft(ws: LayerWorkspace, id: string): DraftLayer {
+// ── Ownership / visibility ───────────────────────────────────────────────
+//
+// Access matrix (#1030). A record's `owner` is the creating principal;
+// undefined (unauthenticated/local) means anyone:
+//
+//   - drafts: session-private storage; mutations (draft_apply_ops,
+//     publish_layer) are owner-only on top of that.
```
[the remark this answers is not in the document]
Comment (louistrue) [reply packages/mcp/src/tools/layer-store.ts]: Fixed in the follow-up commit: the ownership/visibility gating (access matrix, `require*` lookups, `visible*Ids`, unknown-id errors) moved to a new `layer-access.ts`; `layer-store.ts` is back to 352 lines and holds only storage + seeding. All 76 mcp tests and typecheck pass.

**Comment by louistrue** [comment]
@codex review
